app/parse.py

```diff
@@ -15,8 +33,126 @@ class Product:
     num_of_reviews: int
 
 
+TITLE_ROW = [field.name for field in fields(Product)]
+
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="[%(levelname)8s]: %(message)s",
+    handlers=[
+        logging.FileHandler("parser.log"),
+        logging.StreamHandler(sys.stdout)
+    ],
+)
+
+
+class SingletonMeta(type):
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs) -> Any:
+        if cls not in cls._instances:
+            instance = super().__call__(*args, **kwargs)
+            cls._instances[cls] = instance
+        return cls._instances[cls]
+
+
+class SeleniumBrowser(metaclass=SingletonMeta):
+    def __init__(self) -> None:
+        logging.info("Initializing driver")
+        chrome_options = Options()
+        chrome_options.add_argument("--headless")
+        chrome_options.add_argument("--disable-gpu")
+        chrome_options.add_argument("--no-sandbox")
+        chrome_options.add_argument("--disable-dev-shm-usage")
+        self.driver = webdriver.Chrome(options=chrome_options)
+
+    def get_driver(self) -> webdriver:
+        logging.info("Getting driver")
+        return self.driver
+
+    def close_driver(self) -> None:
+        logging.info("Closing driver")
+        self.driver.close()
+        del SeleniumBrowser._instances[type(self)]
+
+
+def parse_single_product(product_soup: Tag) -> Product:
+    return Product(
+        title=product_soup.select_one("a.title")["title"],
+        description=(product_soup.select_one(".description")
+                     .text.replace("\xa0", " ")),
+        price=float(product_soup.select_one(".price").text.replace("$", "")),
+        rating=int(len(product_soup.select("span.ws-icon-star"))),
+        num_of_reviews=int(
+            product_soup.select_one("p.review-count")
+            .text.replace(" reviews", "")
+        ),
+    )
+
+
+def get_products(driver: webdriver) -> List[Product]:
+    page_soup = BeautifulSoup(driver.page_source, "html.parser")
+    products = page_soup.select(".product-wrapper")
+    return [parse_single_product(product) for product in products]
+
+
+def get_products_from_page(page_url: str) -> List[Product]:
+    driver = SeleniumBrowser().get_driver()
+    driver.get(page_url)
+    logging.info(f"Getting products from page {page_url}")
+    time.sleep(5)
```

Comment:
Using `time.sleep(5)` is not the most efficient way to wait for page elements to load. Consider using `WebDriverWait` with appropriate conditions to wait for specific elements to be present or visible, which would make the code more reliable and faster.